### ISSUE TITLE
Custom field layout not saved for custom link type and not selectable

### ISSUE THREAD

**Issue by WHITE-developer** [verbb/hyper#81]
### Describe the bug

I've tried to add a custom link type to be able to select a Formie form, with the idea to be able to select it inside a hyper field and get back to form element or handle to render the form in a modal.

When I add [this](https://verbb.io/craft-plugins/hyper/docs/developers/link-type#element-link-type) to a custom module the settings are shown in the settings field page, but when I change anything to the field layout it isn't getting saved while no error is getting thrown.

The option is also not visble to choose from on field input.

```
<?php

namespace modules\megastoffen\links;

use Craft;
use verbb\formie\elements\Form;
use verbb\hyper\base\ElementLink;
use verbb\hyper\fieldlayoutelements\LinkField;
use verbb\hyper\fields\HyperField;

class Formie extends ElementLink
{
    public static function elementType(): string
    {
        return Form::class;
    }

    public function getSettingsHtml(): ?string
    {
        $variables = $this->getSettingsHtmlVariables();

        return Craft::$app->getView()->renderTemplate("hyper/links/_element/settings", $variables);
    }
    
    public function getInputHtml(LinkField $layoutField, HyperField $field): ?string
    {
        $variables = $this->getInputHtmlVariables($layoutField, $field);

        return Craft::$app->getView()->renderTemplate("hyper/links/_element/input", $variables);
    }
}
```

### Steps to reproduce

1. add a custom link type
2. try to edit or select it

### Craft CMS version

4.4.16.1

### Plugin version

1.1.10

**Comment by engram-design** [comment]
That seems to be working for me (your exact code)!

<img width="779" alt="image" src="https://github.com/verbb/hyper/assets/1221575/e96522fb-098e-450e-ae7a-26eb0112ed77">
<img width="712" alt="image" src="https://github.com/verbb/hyper/assets/1221575/738494b8-ce96-4eec-bec6-7dd9cf6cd4a8">

However, one thing I've noticed is that I needed to add a check for element `uri`. By default, when picking an element in an element link type, only elements with a `uri` are shown as pickable. The thinking was that is something doesn't have a URL, then what's the point in creating a link to it - it won't _have_ a link.

That's still the case, but in your instance, if you want to bypass that and do it anyway, you can include `checkElementUri()`
` (I've updated the docs to reflect that), and be on the `craft-4` branch.

**Comment by WHITE-developer** [comment]
:thinking: 
That's strange for me it seems like the `layoutUid` isn't getting set. The `layoutUid` does is getting set inside the project config file with the right values
Also weird that the styling is of for me but does look to come from the hyper.css file  `.lightswitch.extra-small.on .lightswitch-container .handle{left:calc(50% - 6px)}`
![image](https://github.com/verbb/hyper/assets/22771556/b11c8a0b-33ad-4b92-b301-3fa13643fb82)

**Comment by engram-design** [comment]
So that should certainly be something like:

```
<input type="hidden" name="types[verbb\hyper\fields\HyperField][linkTypes][default-modules-sitemodule-formie][layoutUid]" value="d0f3cbfa-35df-496c-a3bf-fe58c3e878c7">
```

Out of interest, if you add a field to the field layout, click somewhere else (just the background of the field layout builder), and save - does that make a difference?

And yep, there was a change in Craft 4.4.16+ that messed up lightswitches, that'll be fixed shortly.

**Comment by WHITE-developer** [comment]
No doesn't change, the `layoutConfig` input field does get saved changed but the `layoutUid` stays empty

**Comment by WHITE-developer** [comment]
Looks like it goes wrong [here](https://github.com/verbb/hyper/blob/craft-4/src/fields/HyperField.php#L570), every time I reload the field settings page the `$hasLinkType` is empty so it completely recreates the link Type with an empty layout config and UID
Ass it's empty [here](https://github.com/verbb/hyper/blob/craft-4/src/base/Link.php#L197).

If I register the field type directly inside the plugin (https://github.com/verbb/hyper/blob/craft-4/src/services/Links.php#L55) than it does work

**Comment by engram-design** [comment]
Might be worth trying my latest commit which is related to setting link type settings in some instances. To get this early run `composer require verbb/hyper:"dev-craft-4 as 1.1.10"`

It may be worth checking [here](https://github.com/verbb/hyper/blob/ec68bd7bab68a669798350235b043ab1e731fc02/src/fields/HyperField.php#L420) which is triggered when the field settings are saved, and the link types are set.

**Comment by WHITE-developer** [comment]
Looks like the problem was that I called the `EVENT_REGISTER_LINK_TYPES` inside the `Craft::$app->onInit()` function inside my Module.
This prevent the registration happening [here](https://github.com/verbb/hyper/blob/craft-4/src/fields/HyperField.php#L421)

**Comment by engram-design** [comment]
So that _should_ be the correct place to put it. There's an issue https://github.com/verbb/hyper/issues/72 being addressed to check that this function acts properly if run before Craft/Hyper is ready, which might be the issue here.

This should be fixed in https://github.com/verbb/hyper/commit/ec68bd7bab68a669798350235b043ab1e731fc02 for the next release.

**Comment by WHITE-developer** [comment]
Already on the latest commit tho..

**Comment by engram-design** [comment]
Right, I couldn't seem to replicate it, even when using:

```php
Craft::$app->onInit(function() {
    Event::on(Links::class, Links::EVENT_REGISTER_LINK_TYPES, function(RegisterComponentTypesEvent $event) {
        $event->types[] = Formie::class;
    });
});
```

**Comment by engram-design** [comment]
Closing unless this is still an issue?

**Comment by mufi-siebird** [comment]
@engram-design I've created the same with this code:

```
<?php
namespace modules\mensHelperModule\hyper;

use Craft;
use verbb\formie\elements\Form;
use verbb\hyper\base\ElementLink;
use verbb\hyper\fieldlayoutelements\LinkField;
use verbb\hyper\fields\HyperField;

class Formie extends ElementLink
{
    public static function elementType(): string
    {
        return Form::class;
    }

    public function getSettingsHtml(): ?string
    {
        $variables = $this->getSettingsHtmlVariables();

        return Craft::$app->getView()->renderTemplate("hyper/links/_element/settings", $variables);
    }

    public function getInputHtml(LinkField $layoutField, HyperField $field): ?string
    {
        $variables = $this->getInputHtmlVariables($layoutField, $field);

        return Craft::$app->getView()->renderTemplate("hyper/links/_element/input", $variables);
    }

    public static function checkElementUri(): bool
    {
        return false;
    }
}
```

I am inside Neo > Matrix > Hyper LINK

that works and gives me forms, i can select from those form as well. However, I am not able to save the entry. It throws error.

### Possible Solution
If I add a function of URL, it starts working:
```
public function getLinkUrl(): ?string
{
    return '#';
}
```

### Plugins:
Craft CMS: 4.7.1
Hyper: 1.1.19
Neo: 4.0.3

![Screenshot at Jan 30 12-22-33](https://github.com/verbb/hyper/assets/47181126/ed6432da-60b4-4742-b17e-497ad36e1b34)

**Comment by engram-design** [comment]
That's a different issue, but fixed for the next release. To get this early, run `composer require verbb/hyper:"dev-craft-4 as 1.1.19"`.

**Comment by engram-design** [comment]
Fixed in [1.1.20](https://github.com/verbb/hyper/releases/tag/1.1.20)
